wgpu-core/src/device/queue.rs

```
@@ -1279,6 +1285,12 @@ impl Queue {
                                     .unwrap()
                             };
                         }
+                        // It's okay if the texture was destroyed if we still
+                        // have the "pending deletion" handle to it.
+                        Err(DestroyedResourceError(_))
+                            if pending_writes
+                                .temp_textures
+                                .contains_key(&texture.tracker_index()) => {}
                         Err(e) => break 'error Err(e.into()),
```

> It's possible that any texture used in `pending_writes` must either be active or in the list for pending deletion, but I'm not sure if that's the case. If it is, we might want to do something like make the check here a debug-only assertion.

It is the case that all `dst_buffers`/`dst_textures` in pending writes will not be destroyed until the first submission they will be part of completes. Having said that, is it worth adding the new `temp_textures` just for the debug assert here? If so, we should probably do it for buffers too. If not, ignoring the error would get things running.

Note that `.destroy()` on textures wrapping surface textures doesn't do anything.

https://github.com/gfx-rs/wgpu/blob/e936c065fc1bc93af540d9a99981597088a8f2c8/wgpu-core/src/resource.rs#L1187-L1189

---

> Having said that, is it worth adding the new temp_textures just for the debug assert here?

Do you have an opinion on whether this is valuable? The utility of a debug assertion may be limited, if it's not going to be enabled in most Firefox builds. If it is only going to be a debug assertion, possibly we can forego the separate hashmap for `temp_textures`, and do a linear iteration over `temp_resources`.

> Note that .destroy() on textures wrapping surface textures doesn't do anything.

I saw that, but I'm not sure I understand the significance. Not destroying the texture immediately is fine. But if there's other code somewhere else that might independently destroy it before submission, maybe that's a problem?

---

I don't think a debug assertion is necessary in this case. The value of a debug assertion in this case would be to catch erroneous changes where we would destroy the texture prior to the submission. Given that `temp_resources`'s existence is to hold the resources alive and it gets placed in the command buffer I think it's unlikely that reviewers will not scrutinize changes to this area of the codebase.

---

> > Note that .destroy() on textures wrapping surface textures doesn't do anything.
> 
> I saw that, but I'm not sure I understand the significance. Not destroying the texture immediately is fine. But if there's other code somewhere else that might independently destroy it before submission, maybe that's a problem?

I mentioned it since as it stands, the code in `Queue.submit` assumes surface textures are alive, if we were to just replace `Err(e) => break 'error Err(e.into()),` with `Err(e) => {}`, that requires surface textures to never be destroyed (which is the case).